src/psyclone/psyir/backend/fortran.py

```diff
@@ -1157,6 +1157,23 @@ def routine_node(self, node):
             result = f"{self._nindent}program {node.name}\n"
             routine_type = "program"
         else:
+            # Find RoutineSymbol
+            container = node.ancestor(Container)
+            rsym = None
+            if container:
+                rsym = container.symbol_table.get_symbols().get(
```

Comment:
Not for this PR but this idiom makes me think it would be nice if `SymbolTable.lookup()` let you say that a result of `None` was acceptable, rather than having to do do a `try...except` for a KeyError.

Reply:
I feel like it would be better if `lookup` returned `None` always instead of `KeyError` and the `try...except` pattern could just be `if result is None` (or similar pythonic implementation), though I imagine thats probably a decent amount of code needing rewriting.

Reply:
I think I agree with that although it would be a very big change! What do @sergisiso and @hiker think?

Reply:
The KeyError was originally to imitate the python dict interface. We could also add a default value like get has to avoid the KeyError.

But in this case I think we should use the one associated with the Routine which is: self.symbol_table.lookup_with_tag("own_routine_symbol", scope_limit=self)
and will be easily accessible as node.routine_symbol with @JulienRemy PR #2570
and if this it doesn't exist we can error.

(Unfortunately routines have 2 symbols and you will need to check that the elemental/pure properties have been properly set in both - this generates a few issues like if one symbol is updated the other is not. I think we should get rid of the "own_routine_symbol" and instead have a Routine.__symbol - and it's self.name be Routine._symbol.name but this is for a separate issue as it brings some other problems)

Reply:
Is that what #2570 is about? I assumed that the symbol associated with the Routine was for `FUNCTION` and is the symbol associated with their return value and is not the `RoutineSymbol` stored in the scope of the `Container` that contains the `Routine` node. As far as I can work out from the PR/fparser that is how it works, and the only constructor of a `RoutineSymbol` I can see is in `_process_routine_symbols`, which is called from the container or module_handlers.

Also whenever a `RoutineSymbol` is found inside a `Routine` during fparser it is removed and replaced with a `DataSymbol` as far as I can work out - what have I missed here?

Reply:
Is there a reason that a Routine has both its own copy of the RoutineSymbol while its parent also has one? Since searching for the symbol should search above it too?

Reply:
From experimentation this is very difficult behaviour to rely on.

I was correct in part - if you have a function (e.g. `real function x`) then `lookup_with_tag("own_routine_symbol") gives you a `x: DataSymbol<Scalar<REAL, UNDEFINED>, Automatic>`

If you have a subroutine (e.g. `subroutine sub`) then instead you get `sub: RoutineSymbol<NoType, pure=unknown, elemental=unknown>` - this behaviour seems strange to me and I'm not clear why the routine stores a copy of its own routine symbol?

Reply:
You are correct though Sergi - the `is_elemental` and `is_pure` properties are not making it to the "own_routine_symbol" at the moment.

Reply:
Aha its setup in `.name` in `Routine`...

Reply:
Ok if we rename a Routine using `name` all kinds of things could be wrong... This test doesn't pass
```
def test_dummy(fortran_reader):
    code = '''module test
    contains
    impure elemental subroutine sub()
    end subroutine sub
    end module test'''
    container = fortran_reader.psyir_from_source(code)
    module = container.children[0]
    routine = container.children[0].children[0]
    routine.name = "RENAME"
    assert module.symbol_table.lookup(routine.name).name == routine.name
```

Reply:
I can ensure the properties are copied downwards by changing `name` to:
```
            rsymbol = None
            if self._parent:
                try:
                    rsymbol = self.parent.symbol_table.lookup(new_name).copy()                                                                                   except KeyError:
                    pass
            if not rsymbol:
                rsymbol = RoutineSymbol(new_name, NoType())

            self._name = new_name
            # Since the constructor can not mark methods as functions directly
            # the symbol will always start being NoType and must be updated
            # if a return_value type is provided.
            self.symbol_table.add(rsymbol,
                                  tag='own_routine_symbol')
```

I'm not going to change the name change upwards stuff because thats a new issue not related to this I think

Reply:
I created #2582 to solve the upwards change that we should probably have, for after the FileContainer changes also.

Reply:
Thanks both. A lot of this trouble is due to (or summarised by) #1200. Should I therefore review @JulienRemy 's #2570 first do you think?

Reply:
> The KeyError was originally to imitate the python dict interface. We could also add a default value like get has to avoid the KeyError.

This has the advantage that it preserves the existing API so we could make this change incrementally.

Reply:
> I think I agree with that although it would be a very big change! What do @sergisiso and @hiker think?

I'd love this, I hate the constant `try` construct required.

Reply:
> Thanks both. A lot of this trouble is due to (or summarised by) #1200. Should I therefore review @JulienRemy 's #2570 first do you think?

I think this and #2570 are essentially independent so there is no issues with both of these PR being submitted in any order. I think there are separate issues in his PR about whether he is always expecting to get a `RoutineSymbol` (as the #1200 issue highlights). If he always expects/wants a `RoutineSymbol` (and is happy to look in the `Container` to get it if there isn't one in the `Routine`), then his PR probably will need to wait on #2575 , which in turn is waiting on this PR since I know it will affect an xfailing test in this PR.


> This has the advantage that it preserves the existing API so we could make this change incrementally.
I think that probably that is a better initial implementation - but we should perhaps try to have a rule that if someone touches a file with the `KeyError` pattern then the reviewer asks them to change to the new version, and then we can remove it entirely later with less effort?

Reply:
I was wondering why we allow a `Routine` to be renamed and have reminded myself that we do this for Kernels (in PSyKAl) when we transform them (psyGen.py::rename_and_write). Also, in gocean1p0.py::_read_from_device_routine we create a Routine by parsing some generic Fortran and then we rename it. There are probably other instances too. Basically, we can't just turn this off (as I was tempted to argue for).

Reply:
Anyway, all that discussion aside, please could you add a comment explaining your new addition to the `Routine.name` method.